### PR TITLE
add limit of array append

### PR DESCRIPTION
**Pull request by tanZiWen**

### CODE REVIEW

File: vm/neovm/func_validate.go [hunk 1]
```diff
@@ -594,6 +597,52 @@ func validatorReverse(e *ExecutionEngine) error {
...
+			return false
+		}
+		if ok := checkStackSize(v, &i);!ok {
+			return false
+		}
+	}
+	return true
+}
+
+func checkStackSize(s types.StackItems, length *int) bool {
```
Comment (laizy) [inline vm/neovm/func_validate.go]: what are you doing here?

File: vm/neovm/params.go [hunk 2]
```diff
@@ -19,9 +19,9 @@
 package neovm
 
 const (
-	STACK_LIMIT               uint32 = 2 * 1024
+	STACK_LIMIT               int    = 2 * 1024
 	MAX_INVOCATION_STACK_SIZE        = 1024
 	MAX_SIZE_FOR_BIGINTEGER          = 32
 	MAX_BYTEARRAY_SIZE        uint32 = 1024 * 1024
-	MAX_ARRAY_SIZE            uint32 = 1024
+	MAX_ARRAY_SIZE            int    = 1024
```
Comment (laizy) [inline vm/neovm/params.go]: remove `int`, `uint32`,  constant number do not need type info

File: vm/neovm/func_stack_test.go [hunk 3]
```diff
@@ -366,3 +367,16 @@ func TestOpTuck(t *testing.T) {
 		t.Fatal("NeoVM OpTuck test failed.")
 	}
 }
+
+func TestStruct(t *testing.T) {
+	s := types.NewStruct(nil)
+	k := types.NewStruct([]types.StackItems{types.NewInteger(big.NewInt(1))})
+	for i := 0; i < 1024; i++ {
```
Comment (laizy) [inline vm/neovm/func_stack_test.go:374]: from your implementation. i think `1` is enough to lead the following assert to false.

File: vm/neovm/func_validate.go (hunk 1 above)
Comment (tanZiWen) [reply vm/neovm/func_validate.go]: add stack deep check

File: vm/neovm/func_array.go [hunk 4]
```diff
@@ -192,27 +193,27 @@ func opRemove(e *ExecutionEngine) (VMState, error) {
...
 		m, err := item.GetArray()
 		if err != nil {
-			return FAULT, errors.NewErr("[opRemove]get Array error!")
+			return FAULT, fmt.Errorf("%s", "[opRemove]get Array error!")
 		}
 
 		i, err := index.GetBigInteger()
 		if err != nil {
-			return FAULT, errors.NewErr("[opRemove] index not a interger!")
+			return FAULT, fmt.Errorf("%s", "[opRemove] index not a interger!")
```
Comment (WyattMufson) [inline vm/neovm/func_array.go]: Should fix the spelling of "integer"

File: vm/neovm/func_array.go [hunk 5]
```diff
@@ -192,27 +193,27 @@ func opRemove(e *ExecutionEngine) (VMState, error) {
 	case *types.Array:
 		m, err := item.GetArray()
 		if err != nil {
-			return FAULT, errors.NewErr("[opRemove]get Array error!")
+			return FAULT, fmt.Errorf("%s", "[opRemove]get Array error!")
```
Comment (WyattMufson) [inline vm/neovm/func_array.go]: Should add a space after the "]"

**Review by WyattMufson — CHANGES_REQUESTED** [review]

**Comment by WyattMufson** [comment]
This will resolve #930

**Comment by laizy** [comment]
ping @tanZiWen

File: vm/neovm/func_array.go (hunk 4 above)
Comment (tanZiWen) [reply vm/neovm/func_array.go]: yes, it's error spell

File: vm/neovm/func_array.go (hunk 5 above)
Comment (tanZiWen) [reply vm/neovm/func_array.go]: updated

**Review by WyattMufson — DISMISSED** [review]

File: vm/neovm/func_array.go (hunk 4 above)
Comment (WyattMufson) [reply vm/neovm/func_array.go]: Great!

File: vm/neovm/func_array.go (hunk 5 above)
Comment (WyattMufson) [reply vm/neovm/func_array.go]: Awesome

File: vm/neovm/func_stack_test.go [hunk 6]
```diff
@@ -366,3 +367,16 @@ func TestOpTuck(t *testing.T) {
...
+		k = types.NewStruct([]types.StackItems{k})
+	}
+
+	s.Add(k)
+	s.Add(s)
+
+	assert.False(t, CheckStackSize([]types.StackItems{s, s, s}))
```
Comment (laizy) [inline vm/neovm/func_stack_test.go:381]: check 1023 is ok.

File: vm/neovm/func_validate.go [hunk 7]
```diff
@@ -594,6 +597,52 @@ func validatorReverse(e *ExecutionEngine) error {
...
 }
 
+func CheckArraySize(count int) bool {
+	if count > MAX_ARRAY_SIZE {
+		return false
+	}
+	return true
+}
+
+func CheckStackSize(items []types.StackItems) bool {
```
Comment (laizy) [inline vm/neovm/func_validate.go]: remove this

Comment (tanZiWen) [reply vm/neovm/func_validate.go]: done

**Comment by laizy** [comment]
fixed in #958

**Pull request closed without merging**